server/src/test/java/org/elasticsearch/cluster/routing/allocation/ThrottlingAllocationDeciderTests.java

```diff
@@ -0,0 +1,217 @@
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.shard.ShardId;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class ThrottlingAllocationDeciderTests extends ESAllocationTestCase {
+
+    private record TestHarness(
+        ClusterState clusterState,
+        RoutingNodes mutableRoutingNodes,
+        RoutingNode mutableRoutingNode1,
+        RoutingNode mutableRoutingNode2,
+        ShardRouting shardRouting1Primary,
+        ShardRouting shardRouting1Replica,
+        ShardRouting shardRouting2Primary,
+        ShardRouting shardRouting2Replica
+    ) {}
+
+    private TestHarness setUpTwoNodesAndIndexWithTwoUnassignedPrimariesAndReplicas() {
+        int numberOfShards = 2;
+        ClusterState clusterState = ClusterStateCreationUtils.stateWithUnassignedPrimariesAndReplicas(
+            new String[] { "test-index" },
+            numberOfShards,
+            1
+        );
+        // The number of data nodes the util method above creates is numberOfReplicas+2.
+        assertEquals(3, clusterState.nodes().size());
+        assertEquals(1, clusterState.metadata().getTotalNumberOfIndices());
+
+        var indexIterator = clusterState.metadata().indicesAllProjects().iterator();
+        assertTrue(indexIterator.hasNext());
+        IndexMetadata testIndexMetadata = indexIterator.next();
+        assertFalse(indexIterator.hasNext());
+        Index testIndex = testIndexMetadata.getIndex();
+        assertEquals(numberOfShards, testIndexMetadata.getNumberOfShards());
+        ShardId testShardId1 = new ShardId(testIndex, 0);
+        ShardId testShardId2 = new ShardId(testIndex, 1);
+
+        var mutableRoutingNodes = clusterState.mutableRoutingNodes();
+
+        // The RoutingNode references must be to the RoutingAllocation's RoutingNodes, so that changes to one is reflected in the other.
+        var routingNodesIterator = mutableRoutingNodes.iterator();
+        assertTrue(routingNodesIterator.hasNext());
+        var mutableRoutingNode1 = routingNodesIterator.next();
+        assertTrue(routingNodesIterator.hasNext());
+        var mutableRoutingNode2 = routingNodesIterator.next();
+        assertTrue(routingNodesIterator.hasNext());
+        routingNodesIterator.next(); // unused third node.
+        assertFalse(routingNodesIterator.hasNext());
+
+        ShardRouting shardRouting1Primary = TestShardRouting.newShardRouting(testShardId1, null, null, true, ShardRoutingState.UNASSIGNED);
+        ShardRouting shardRouting2Primary = TestShardRouting.newShardRouting(testShardId2, null, null, true, ShardRoutingState.UNASSIGNED);
+        ShardRouting shardRouting1Replica = TestShardRouting.newShardRouting(testShardId1, null, null, false, ShardRoutingState.UNASSIGNED);
+        ShardRouting shardRouting2Replica = TestShardRouting.newShardRouting(testShardId2, null, null, false, ShardRoutingState.UNASSIGNED);
```

Comment:
Maybe it's not important, but I think we should be able to pull these out of the routing table with

```java
        RoutingTable routingTable = clusterState.routingTable(ProjectId.DEFAULT);
        routingTable.shardRoutingTable(shardId).primaryShard();
        routingTable.shardRoutingTable(shardId).replicaShards().get(0);
```

Then perhaps we can avoid the change to `TestShardRouting`?